commatrix/generate.go

```diff
@@ -0,0 +1,207 @@
+	"github.com/openshift-kni/commatrix/ss"
+	"github.com/openshift-kni/commatrix/types"
+)
+
+func GenerateSS(kubeconfig, customEntriesPath, customEntriesFormat, format string, env Env, deployment Deployment, destDir string) (ssMat *types.ComMatrix, ssOutTCP, ssOutUDP []byte, err error) {
+	cs, err := clientutil.New(kubeconfig)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	nodesList, err := cs.CoreV1Interface.Nodes().List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	nodesComDetails := []types.ComDetails{}
+	err = debug.CreateNamespace(cs, consts.DefaultDebugNamespace)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+	defer func() {
+		err := debug.DeleteNamespace(cs, consts.DefaultDebugNamespace)
+		if err != nil {
+			fmt.Printf("failed to delete debug namespace: %v", err)
+		}
+	}()
+
+	nLock := &sync.Mutex{}
+	g := new(errgroup.Group)
+	for _, n := range nodesList.Items {
+		node := n
+		g.Go(func() error {
+			debugPod, err := debug.New(cs, node.Name, consts.DefaultDebugNamespace, consts.DefaultDebugPodImage)
+			if err != nil {
+				return err
+			}
+			defer func() {
+				err := debugPod.Clean()
+				if err != nil {
+					fmt.Printf("failed cleaning debug pod %s: %v", debugPod, err)
+				}
+			}()
+
+			cds, ssTCP, ssUDP, err := ss.CreateSSOutputFromNode(debugPod, &node)
+			if err != nil {
+				return err
+			}
+			nLock.Lock()
+			ssTCPLine := fmt.Sprintf("node: %s\n%s\n", node.Name, string(ssTCP))
+			ssUDPLine := fmt.Sprintf("node: %s\n%s\n", node.Name, string(ssUDP))
+
+			nodesComDetails = append(nodesComDetails, cds...)
+			ssOutTCP = append(ssOutTCP, []byte(ssTCPLine)...)
+			ssOutUDP = append(ssOutUDP, []byte(ssUDPLine)...)
+			nLock.Unlock()
+			return nil
+		})
+	}
+
+	err = g.Wait()
+	if err != nil {
+		return nil, nil, nil, err
+	}
+	cleanedComDetails := types.CleanComDetails(nodesComDetails)
+	ssComMat := types.ComMatrix{Matrix: cleanedComDetails}
+	return &ssComMat, ssOutTCP, ssOutUDP, nil
+}
+
+func getPrintFunction(format string) (func(m types.ComMatrix) ([]byte, error), error) {
+	switch format {
+	case "json":
+		return types.ToJSON, nil
+	case "csv":
+		return types.ToCSV, nil
+	case "yaml":
+		return types.ToYAML, nil
+	case "nft":
+		return types.ToNFTables, nil
+	default:
+		return nil, fmt.Errorf("invalid format: %s. Please specify json, csv, yaml, or nft", format)
+	}
+}
+
+func createSSOutputFiles(destDir string) (*os.File, *os.File, error) {
+	tcpFile, err := os.OpenFile(path.Join(destDir, "raw-ss-tcp"), os.O_TRUNC|os.O_CREATE|os.O_WRONLY, 0644)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	udpFile, err := os.OpenFile(path.Join(destDir, "raw-ss-udp"), os.O_TRUNC|os.O_CREATE|os.O_WRONLY, 0644)
+	if err != nil {
+		tcpFile.Close()
+		return nil, nil, err
+	}
+
+	return tcpFile, udpFile, nil
+}
+
+func WriteSSRawFiles(destDir string, ssOutTCP, ssOutUDP []byte) error {
+	tcpFile, udpFile, err := createSSOutputFiles(destDir)
```

Comment:
I still think createSSOutputFiles is not needed and confusing, it's not really creating them, since the generate func is the one creating them, it's only opening the files